### ISSUE TITLE
Travis FSLeyes installs failing

### ISSUE THREAD

**Issue by mariehbourget** [neuropoly/axondeepseg#390]
For the last 2 days, all Travis builds with FSLeyes install are failing on `test_train_network.py`.

The error comes from Keras and happened in `train_network.py` where loading the trained model.
<details>
  <summary>Error ADS train_network.py</summary>
  
```
test/test_train_network.py:135: 
AxonDeepSeg/train_network.py:289: in train_model
    str(path_model) + "/model.hdf5", custom_objects=custom_objects
../../../miniconda/envs/ads_venv/lib/python3.7/site-packages/keras/engine/saving.py:419: in load_model
    model = _deserialize_model(f, custom_objects, compile)
```
</details>

<details>
  <summary>Error Keras</summary>
  
```
        model_config = f['model_config']
        if model_config is None:
            raise ValueError('No model found in config.')
>       model_config = json.loads(model_config.decode('utf-8'))
E       AttributeError: 'str' object has no attribute 'decode'
../../../miniconda/envs/ads_venv/lib/python3.7/site-packages/keras/engine/saving.py:224: AttributeError
```
</details>

The same test with Basic Install is passing.

This may be linked with the issue we had with `h5py` version in PR #382.

The "basic install" installs `h5py=2.10.0`.
The "FSLeyes install" with conda-forge seems to install `h5py=3.1.0`.

The last build with FSLeyes that passed was 5 days ago, and FSLeyes was installing `h5py=2.10.0`.

**Comment by mariehbourget** [comment]
I tested the installation locally in a new conda environment.
All tests passed after ADS install.
`test_train_network.py` fails with the same error after FSLeyes install (Linux)
But `conda list` gives the version `h5py=2.10.0` after both installation, so it may not be related after all?

In addition, `test_floodfill_axons_returns_expected_arrays` from `test_postprocessing` fails with:
<details>
  <summary>Error</summary>
 
``` 
_______________________________ TestCore.test_floodfill_axons_returns_expected_arrays _______________________________

self = <test_postprocessing.TestCore object at 0x7f8d66f46450>

    @pytest.mark.unit
    def test_floodfill_axons_returns_expected_arrays(self):
        before_ff_axon_mask, before_ff_myelin_mask = \
            ads_utils.extract_axon_and_myelin_masks_from_image_data(self.before_floodfill_image)
        after_ff_expected_axon_mask, _ = \
            ads_utils.extract_axon_and_myelin_masks_from_image_data(self.after_floodfill_image)
        after_ff_obtained_axon_mask = postprocessing.floodfill_axons(before_ff_axon_mask, before_ff_myelin_mask)

>       assert np.array_equal(after_ff_obtained_axon_mask, after_ff_expected_axon_mask)
E       assert False
E        +  where False = <function array_equal at 0x7f8d924daa70>(array([[1, 1, 1, ..., 1, 1, 1],\n       [1, 1, 1, ..., 1, 1, 1],\n       [1, 1, 1, ..., 1, 1, 1],\n       ...,\n       [1, 1, 1, ..., 1, 1, 1],\n       [1, 1, 1, ..., 1, 1, 1],\n       [1, 1, 1, ..., 1, 1, 1]], dtype=uint8), array([[1, 1, 1, ..., 0, 0, 0],\n       [1, 1, 1, ..., 0, 0, 0],\n       [1, 1, 1, ..., 0, 0, 0],\n       ...,\n       [0, 0, 0, ..., 0, 0, 0],\n       [0, 0, 0, ..., 0, 0, 0],\n       [0, 0, 0, ..., 0, 0, 0]], dtype=uint8))
E        +    where <function array_equal at 0x7f8d924daa70> = np.array_equal

test/test_postprocessing.py:40: AssertionError
 ``` 
</details>

@mathieuboudreau do you have an idea?

**Comment by mathieuboudreau** [comment]
Hmmm. I know that installing FSLeyes using conda overrides some of the requirement versions of our original ADS install sometimes. Could you try pip installing ADS, installing FSLeyes, then pip installing ADS again to see if that it resolves? If so, we could do a comparison of the `pip freeze` output before and after FSLeyes being installed.

**Comment by mariehbourget** [comment]
Re-installing ADS after the FSLeyes install resolved the error on `test_train_network.py`.

The list of packages downgraded is  at the end of the ADS re-install verbose (along with an error message about `fslpy`)
<details>
  <summary>ADS re-install message</summary>
 
``` 
Installing collected packages: numpy, scipy, matplotlib, idna, urllib3, requests, h5py, AxonDeepSeg
  Attempting uninstall: numpy
    Found existing installation: numpy 1.19.4
    Uninstalling numpy-1.19.4:
      Successfully uninstalled numpy-1.19.4
  Attempting uninstall: scipy
    Found existing installation: scipy 1.5.3
    Uninstalling scipy-1.5.3:
      Successfully uninstalled scipy-1.5.3
  Attempting uninstall: matplotlib
    Found existing installation: matplotlib 3.1.1
    Uninstalling matplotlib-3.1.1:
      Successfully uninstalled matplotlib-3.1.1
  Attempting uninstall: idna
    Found existing installation: idna 2.10
    Uninstalling idna-2.10:
      Successfully uninstalled idna-2.10
  Attempting uninstall: urllib3
    Found existing installation: urllib3 1.25.11
    Uninstalling urllib3-1.25.11:
      Successfully uninstalled urllib3-1.25.11
  Attempting uninstall: requests
    Found existing installation: requests 2.24.0
    Uninstalling requests-2.24.0:
      Successfully uninstalled requests-2.24.0
  Attempting uninstall: h5py
    Found existing installation: h5py 3.1.0
    Uninstalling h5py-3.1.0:
      Successfully uninstalled h5py-3.1.0
  Attempting uninstall: AxonDeepSeg
    Found existing installation: AxonDeepSeg 3.2.0
    Uninstalling AxonDeepSeg-3.2.0:
      Successfully uninstalled AxonDeepSeg-3.2.0
  Running setup.py develop for AxonDeepSeg
ERROR: After October 2020 you may experience errors when installing or updating packages. This is because pip will change the way that it resolves dependency conflicts.

We recommend you use --use-feature=2020-resolver to test your packages with the new resolver before it becomes the default.

fslpy 3.4.0 requires dataclasses, which is not installed.
Successfully installed AxonDeepSeg h5py-2.10.0 idna-2.7 matplotlib-3.1.0 numpy-1.16.4 requests-2.20.1 scipy-1.3.0 urllib3-1.24.3
```
</details>

This is confirmed by the `diff` on `pip list` comparing the list after the fsleyes install with the list after ADS re-install
<details>
  <summary>diff pip list</summary>
 
``` 
diff fsleyes-piplist.txt re-ads-piplist.txt
37,38c37,38
< h5py                          3.1.0
< idna                          2.10
---
> h5py                          2.10.0
> idna                          2.7
64c64
< matplotlib                    3.1.1
---
> matplotlib                    3.1.0
76c76
< numpy                         1.19.4
---
> numpy                         1.16.4
116c116
< requests                      2.24.0
---
> requests                      2.20.1
120c120
< scipy                         1.5.3
---
> scipy                         1.3.0
145c145
< urllib3                       1.25.11
---
> urllib3                       1.24.3
```
</details>

So `h5py` is indeed overrides by fsleyes install from `2.1.10` to `3.1.0`.

Also, `test_postprocessing.py` is still failing locally, which was not the case on Travis.

**Comment by mathieuboudreau** [comment]
Great! That means we're on the right track.

As for test_postprocessing.py, I think the best thing we could do is to find out what is happening is save (and not delete) the output files from the tests and compare them visually with the test files. I had some minor concerns while reviewing that the tests may be too specific (because we're comparing all poxels of one image with all pixels of another, even a minor difference in flood filling performance might make the test fail), but thought that I'd leave it to a future test failing to look into this further. Maybe @Stoyan-I-A could look into this since he's written those tests?

**Comment by mathieuboudreau** [comment]
And for h5py, we could either contact FSLeyes through github to make them aware of the h5py bug that we've needed to resolve using a fixed lower version both here and in another NeuroPoly project, or add another installation instruction after installing FSLeyes until that one is fixed by h5py folks. FSLeyes likely just has a `>x.x.x` rule, so it's always going to download the latest version (which I dislike, but we have to deal with it nonetheless).

**Comment by mariehbourget** [comment]
I think we should do both:
- contact FSLeyes through github as this will likely be a bug for FSLeyes with SCT as well (could you take care of that part?)
- in the meantime, add another instruction installation to our doc

I can take care of the instructions, I think we should add the `pip install -e .` of axondeepseg after FSLeyes to make sure all the otehr packages are downgraded as well (not only h5py), what do you think?
I already have a PR opened for documentation, should I include it in the PR or open another one?

**Comment by jcohenadad** [comment]
> And for h5py, we could either contact FSLeyes through github to make them aware of the h5py bug that we've needed to resolve using a fixed lower version both here and in another NeuroPoly project,

is that really a bug in FSLeyes, or a bug in Keras/TF that does not support `h5py=3.1.0`? I would think the latter, no?

**Comment by mariehbourget** [comment]
> is that really a bug in FSLeyes, or a bug in Keras/TF that does not support `h5py=3.1.0`? I would think the latter, no?

From my understanding , the bug is in Keras/TF, but we have it in fsleyes because it installs the latest version of `h5py`, different from the fixed version in ADS/SCT (`2.10.0`).

> In terms of potential fix: I agree with running `pip install -e .` afterwards.

I'll do a PR now to fix documentation and travis test.

>I think it would be great to also have a CLI that check dependencies, similar to [SCT's `sct_check_dependencies`]https://github.com/shimming-toolbox/shimming-toolbox/issues/161#issuecomment-725028653).

I agree, I'll open an issue for this new feature.

**Comment by mathieuboudreau** [comment]
Fair point re: h5py!

The problem with doing `pip install -e .` after installing FSLeyes is that we might downgrade other packages are needed for FSLeyes too, breaking that tool now. Maybe just redownloading the proper version h5py would be more prudent.

**Comment by Stoyan-I-A** [comment]
> `test_floodfill_axons_returns_expected_arrays` from `test_postprocessing` fails

I will test this on my end whenever I have time before our next meeting. 

Here's some notes:
- MH had an issue (#343) where the floodfill wouldn't work properly on her end. I couldn't reproduce her results and the issue was fixed by re-installing ADS in a new conda environment
- The test passes on travis
- The FSLeyes installation might change the `Pillow` and/or `NumPy` packages versions.

**Comment by mariehbourget** [comment]
> * MH had an issue (#343) where the floodfill wouldn't work properly on her end. I couldn't reproduce her results and the issue was fixed by re-installing ADS in a new conda environment

Thanks @Stoyan-I-A for reminding me! You are right, it looks like a similar problem and the fill axons in FSLeyes has the same issue that in #343 even with a brand new installation . As it only happens on my end (travis is OK), I'll get back to you with details about installation procedures and packages installed asap (tonight of tomorrow).

**Comment by mariehbourget** [comment]
I moved the discussion about the floodfill test in #393, as it is a different issu.

The issue with travis tests in FSLeyes install is fixed in #392, so I'm closing this issue.
